### DISCUSSION TITLE
where is the code that data node and query node handles the insert and delete which are part of upsert

### DISCUSSION THREAD

**Discussion by ron-gilaad** [milvus-io/milvus#37056] [Q&A and General discussion] [answered]
Hi,
The Execute function of the upsertTask calls stream.Produce(msgPack) to send the DML operations encapsulated in msgPack (which contains the insert and delete operations) to the stream for further processing by data and query nodes. Where is the code that data node and query node handles it?

**Comment by yhmo** [comment, marked as answer]
In v2.4, datanode processes insert messages in this place:
https://github.com/milvus-io/milvus/blob/1d9c7462baf07b6f2c491d675fcdc9a9e12bdcfb/internal/datanode/flow_graph_dd_node.go#L164

querynode processes insert message in this place:
https://github.com/milvus-io/milvus/blob/1d9c7462baf07b6f2c491d675fcdc9a9e12bdcfb/internal/querynodev2/pipeline/insert_node.go#L89


In master branch, datanode processes insert messages in this place:
https://github.com/milvus-io/milvus/blob/5a1f75227249914ef96df70a6acc399fa0c8aa30/internal/flushcommon/pipeline/flow_graph_dd_node.go#L172

**Reply by ron-gilaad** [reply]
thanks @yhmo I found it myself according to the log
